Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/AfterDirectoryInitializationRunner.java

```diff
@@ -0,0 +1,51 @@
+package org.dolphinemu.dolphinemu.utils;
+
+import android.content.Context;
+import android.content.IntentFilter;
+import android.support.v4.content.LocalBroadcastManager;
+
+public class AfterDirectoryInitializationRunner
+{
+  private DirectoryStateReceiver directoryStateReceiver;
+
+  /**
+   * Executes a Runnable after directory initialization has finished.
+   *
+   * If this is called when directory initialization already is done,
+   * the Runnable will be executed immediately. If this is called before
+   * directory initialization is done, the Runnable will be executed
+   * after directory initialization finishes successfully, or never
+   * in case directory initialization doesn't finish successfully.
+   *
+   * If this is called when a previous invocation still is waiting to
+   * run, the previous invocation will be forgotten.
+   */
+  public void run(Context context, Runnable runnable)
+  {
+    if (!DirectoryInitialization.areDolphinDirectoriesReady())
+    {
+      // Wait for directories to get initialized
+      IntentFilter statusIntentFilter = new IntentFilter(
+              DirectoryInitialization.BROADCAST_ACTION);
+
+      directoryStateReceiver = new DirectoryStateReceiver(directoryInitializationState ->
+      {
+        if (directoryInitializationState ==
+                DirectoryInitialization.DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED)
+        {
+          LocalBroadcastManager.getInstance(context).unregisterReceiver(directoryStateReceiver);
+          directoryStateReceiver = null;
+          runnable.run();
+        }
+      });
+      // Registers the DirectoryStateReceiver and its intent filters
+      LocalBroadcastManager.getInstance(context).registerReceiver(
```

Comment:
Is this a 1:1 thing? It feels like it would just be a list of receivers, each notified in turn when something happens.
If it is indeed a list, I'd expect any older invocations to still be alive; rendering the last paragraph of the comment above slightly fuzzy.
Unless we want it to be like that; then we're probably supposed to call `unregisterReceiver` in case `directoryStateReceiver` is not `null`.

...not that it matters too much, since you keep creating new instances in the places where it is used; so they'd likely end up running anyways.

Reply:
I was assuming that the old `directoryStateReceiver` would get deallocated once we create a new `directoryStateReceiver`, but I suppose that might not be true since the `LocalBroadcastManager` keeps a reference to it... But then why does all the previously existing code that uses a `DirectoryStateReceiver` keep it around as a class member?

Since I'm not entirely sure how it works, I think I'll just change the comment to say that multiple calls is not a supported use case, because that's how I've been thinking of it while writing the code.

Reply:
Actually, I'll just delete the part of the comment that mentions multiple calls.

Reply:
I don't know, but if it works anything close to C# wrt the GC, it should still be referenced by the `LocalBroadcastManager` (and the closure itself should still be referencing the passed `runnable`, keeping it alive as well)

Reply:
Disregard my previous comment – I've realized that multiple calls should not be encouraged, because if old `DirectoryStateReceiver`s indeed are kept around (which is likely), the way we do unregistration means that newer calls will get dropped rather than old calls. I will make the comment say that multiple calls is not a supported use case.